crates/metadata-store/src/local/grpc/net_util.rs

```diff
@@ -0,0 +1,238 @@
+}
+
+async fn run_listener_loop<L, S, B>(
+    mut listener: L,
+    service: S,
+    server_name: &'static str,
+) -> Result<(), Error>
+where
+    L: Listener,
+    L::Io: Send + Unpin + 'static,
+    L::Addr: Debug,
+    S: hyper::service::Service<http::Request<Incoming>, Response = hyper::Response<B>>
+        + Send
+        + Clone
+        + 'static,
+    S::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+    S::Future: Send,
+    B: Body + Send + 'static,
+    B::Data: Send,
+    B::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    let mut shutdown = std::pin::pin!(cancellation_watcher());
+    let tc = task_center();
+    let executor = TaskCenterExecutor::new(tc.clone(), server_name);
+    loop {
+        tokio::select! {
+            biased;
+            _ = &mut shutdown => {
+                break;
+            }
+            incoming_connection = listener.accept() => {
+                let (stream, remote_addr) = incoming_connection?;
+                let io = TokioIo::new(stream);
+
+                debug!("Accepting incoming connection from '{remote_addr:?}'.");
+
+                tc.spawn_child(TaskKind::RpcConnection, server_name, None, handle_connection(
+                    io,
+                    service.clone(),
+                    executor.clone(),
+                ))?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn handle_connection<S, B, I>(
+    io: I,
+    service: S,
+    executor: TaskCenterExecutor,
+) -> anyhow::Result<()>
+where
+    S: hyper::service::Service<http::Request<Incoming>, Response = hyper::Response<B>>
+        + Send
+        + Clone
+        + 'static,
+    S::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+    S::Future: Send,
+    B: Body + Send + 'static,
+    B::Data: Send,
+    B::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+    I: Read + Write + Unpin + 'static,
+{
+    let builder = hyper_util::server::conn::auto::Builder::new(executor);
+    let connection = builder.serve_connection(io, service);
+
+    tokio::select! {
+        res = connection => {
+            // propagate errors
+            res.map_err(Error::HandlingConnection)?;
+        },
+        _ = cancellation_watcher() => {}
+    }
+
+    Ok(())
+}
+
+#[derive(Clone)]
+struct TaskCenterExecutor {
+    task_center: TaskCenter,
+    name: &'static str,
+}
+
+impl TaskCenterExecutor {
+    fn new(task_center: TaskCenter, name: &'static str) -> Self {
+        Self { task_center, name }
+    }
+}
+
+impl<F> hyper::rt::Executor<F> for TaskCenterExecutor
+where
+    F: Future + 'static + Send,
+    F::Output: Send + 'static,
+{
+    fn execute(&self, fut: F) {
+        // ignore shutdown error
+        let _ =
+            self.task_center
+                .spawn_child(TaskKind::RpcConnection, self.name, None, async move {
+                    // ignore the future output
+                    let _ = fut.await;
+                    Ok(())
+                });
+    }
+}
```

Comment:
Also this one makes sense to live in its own rust crate.

Reply:
Follow-up issue.